### PR TITLE
add imgaug module

### PR DESCRIPTION
**Pull request by ZhaoChuyang**
PR for GSoC'22 project on Efficient Data Augmentation Module in OpenCV for DL Training

I implemented data augmentation methods based on basic image processing and have tested the performance of these already implemented methods in python enviroment. The following table demonstrates the running time (in seconds) comparison between OpenCV-Aug module and torchvision transforms on a subset of ImageNet:

**single method:**

method | config | dataset | opencv | torchvision
-- | -- | -- | -- | --
resize | size: (200, 200) | imagenet-320 | 0.388 | 1.745
center crop | size: (200, 200) | imagenet-320 | 0.001 | 0.017
pad | padding: (100, 100, 100, 100) | imagenet-320 | 0.141 | 0.553
random crop | size: (200, 200) | imagenet-320 | 0.001 | 0.028
random resized crop | size: (500, 500) | imagenet-320 | 0.368 | 3.806
random flip | default | imagenet-320 | 0.046 | 0.519

**compose multiple methods:**

method | config | num_augs | opencv | pytorch
-- | -- | -- | -- | --
RandomCrop + RandomFlip + Pad | RandomCrop: size(300, 300) Pad: padding(100, 100, 100, 100) | 3 | 0.481 | 1.240
Resize + Pad + RandomFlip + CenterCrop | Resize: size(400, 400) Pad:  padding(100, 100, 100, 100) CenterCrop: size(200, 200) | 5 | 0.636 | 5.486

Besides augmentation methods for pure images, augmentation methods for detection task and segmentation task is also added, which requires processing the target labels of corresponding tasks.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by kaingwade — COMMENTED** [review]
Please add the short-form license and the documentation for the functionalites.

File: modules/aug/samples/opencv_aug_demo.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def main():
+    # read image
+    input_path = "../../../samples/data/corridor.jpg"
+    image = cv2.imread(input_path)
+
+    image = compose(image)
+
+    plt.imshow(image)
```
Comment (kaingwade) [inline modules/aug/samples/opencv_aug_demo.py:46]: Can you call cv2.imshow() here to show the result?

File: modules/aug/src/transforms.cpp [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+//        if(pad_if_need && src.cols < sz.width){
+//            Vec4i _padding = {0, 0, sz.width - src.cols, sz.width - src.cols};
+//            copyMakeBorder(src, src, _padding[0], _padding[1], _padding[2], _padding[3], padding_mode, fill);
+//        }
+//
+//        int x, y;
+//        getRandomCropParams(src.rows, src.cols, sz.height, sz.width, &x, &y);
+//        Mat cropped(src, Rect(x, y, sz.width, sz.height));
+//        (*(Mat*)_src.getObj()) = cropped;
+//    }
```
Comment (kaingwade) [inline modules/aug/src/transforms.cpp:71]: Please remove code which are not finished, experimental, for debug and the like.

**Comment by ZhaoChuyang** [comment]
Hi @kaingwade, thanks for the advice. I have added the license and add documentation for imgaug module. Junk codes in header files have also been cleaned.

File: modules/aug/samples/opencv_aug_demo.py (hunk 1 above)
Comment (ZhaoChuyang) [reply modules/aug/samples/opencv_aug_demo.py]: OK, I will change the python sample in next commit.

**Comment by asenyaev** [comment]
Hello @ZhaoChuyang!

I have added support for `opencv_test_imgaug` in CI. Could you push the latest commit again to re-run CI for this PR?

**Comment by ZhaoChuyang** [comment]
Hi @asenyaev, I have commited the latest changes.

File: modules/imgaug/include/opencv2/imgaug/functional.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    //! @{
+
+    /** @brief Adjust the brightness of the given image.
+     *
+     * @param img Source image. This operation is inplace.
+     * @param brightness_factor brightness factor which controls the brightness of the adjusted image.
+     * Brightness factor should be >= 0. When brightness factor is larger than 1, the output image will be brighter than original.
+     * When brightness factor is less than 1, the output image will be darker than original.
+     */
+    void adjustBrightness(Mat& img, double brightness_factor);
```
Comment (asmorkalov) [inline modules/imgaug/include/opencv2/imgaug/functional.hpp:20]: It makes sense to use `InputArray` and `OutputArray` for all functions in this header. Rationale:
- Generic OpenCV interface.
- UMat support, if `InputArray` is promoted from `Transform` classes directly without getMat() call.

File: modules/imgaug/misc/python/pyopencv_imgaug.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        if(!obj || obj == Py_None)
+            return true;
+        if(PyLong_Check(obj)){
+            value = PyLong_AsUnsignedLongLong(obj);
+        }else{
+            return false;
+        }
+        return value != (unsigned int)-1 || !PyErr_Occurred();
+    }
+};
```
Comment (asmorkalov) [inline modules/imgaug/misc/python/pyopencv_imgaug.hpp:43]: I have not found `long log` or `unsigned long long` usage in interface. Most probably the manual binding is not required.

File: modules/imgaug/src/functional.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        int num_channels = img.channels();
+        if(num_channels != 1 && num_channels != 3){
+            CV_Error(Error::BadNumChannels, "Only support images with 1 or 3 channels");
+        }
+        if(img.channels() == 1) return;
+        Mat gray;
+        cvtColor(img, gray, COLOR_BGR2GRAY);
+        std::vector<Mat> gray_arrays = {gray, gray, gray};
+        merge(gray_arrays, gray);
```
Comment (asmorkalov) [inline modules/imgaug/src/functional.cpp:47]: it chould be cvtColor with COLOR_GRAY2BGR.

File: modules/imgaug/src/functional.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        int num_channels = img.channels();
+        if(num_channels != 1 && num_channels != 3){
+            CV_Error(Error::BadNumChannels, "Only support images with 1 or 3 channels");
+        }
+       Mat* channels = new Mat[num_channels];
+        split(img, channels);
+        std::vector<Mat> new_channels;
+        for(int i=0; i < num_channels; i++){
+            Mat& channel = channels[i];
+            Scalar avg = mean(channel);
```
Comment (asmorkalov) [inline modules/imgaug/src/functional.cpp:27]: The function [cv::mean](https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga191389f8a0e58180bb13a727782cd461) calculates the mean value M of array elements, independently for each channel, and return it. No need to allocate array and iterate over channels the next arithmetic steps could be dome for all channels together without loop.

**Comment by asmorkalov** [comment]
General notes on test code:
- There is no need to store cropped/flipped/converted images in opencv_extra, if test checks single transformation. It's better to call reference function from OpenCV itself. The test should check new logic, but not OpenCV primitives behavior.

**Comment by asmorkalov** [comment]
@ZhaoChuyang Friendly reminder.

**Comment by ZhaoChuyang** [comment]
Hi, sorry for the delay, I have been caught up in a DDL. I'll fix them ASAP.

**Comment by LaurentBerger** [comment]
What's new about this module?
